### DISCUSSION TITLE
Using Kinc with Zig

### DISCUSSION THREAD

**Discussion by bluesillybeard** [Kode/Kore#837] [Q&A] [answered]
Here's the deal: I am wanting to use Kinc in a project of mine. The issue is, I want to use the Zig programming language.

If you aren't familiar with Zig, it's a low-level language that has very good support for interacting with C and C++. It has its own build system, which although is still in development is stable enough that some C projects (such as Raylib) are using it though they don't use Zig itself.

Look at https://ziglang.org for more info on Zig.

Zig's C/C++ compiler uses clang under the hood, in case that matters.

I see a few ways to do this:
- translate the entire kfile.js and add that to my build.zig
    - I would be willing to submit a poll request that adds support for Zig's build system, if enough people are interested
    - this is my favorite option, although it would be a substantial amount of work to create and maintain
- compile static libraries for every platform I want to support and link that with my project
    - This would take quite a while, and updating would become a chore, so this is my least favorite option

I would also like to take advantage of Zig's easy cross-compilation feature, so  I can just run an export script on my Linux machine to create all of the artifacts. But that's not mandatory, setting up a bunch of VMs for exporting is not the end of the world.

Any help is appreciated, thanks!

P.S: I noticed `create_single_header_libs.js`, does that do anything that might be useful to me?

**Comment by RobDangerous** [comment]
Simply add an exporter to kmake that creates a zig build file. See https://github.com/Kode/kmake/tree/main/kmake/src/Exporters
You can use something simple like the makefile-exporter at https://github.com/Kode/kmake/blob/main/kmake/src/Exporters/LinuxExporter.ts#L232 as a guideline.

**Reply by bluesillybeard** [reply]
This sounds like a really good option, thanks!

**Comment by bluesillybeard** [comment]
Looking into it further, I think I am going to use bgfx instead of Kinc.

I can simply use their amalgamated build ([See the `Amalgamated Build` section of this page](https://bkaradzic.github.io/bgfx/build.html)) and include that singular C++ file as part of my build process.

**Reply by RobDangerous** [reply]
Booo!

**Reply by bluesillybeard** [reply]
The amalgamated build of bgfx is a lie. It's actually a massive pain in the butt and completely not worth it.

**Reply by RobDangerous** [reply]
Yaay!

**Comment by bluesillybeard** [comment, marked as answer]
I've created a helper for linking Kinc with Zig.

Here's my [repo for the project](https://github.com/bluesillybeard/Kinc.zig)

However, I've encountered an issue that I don't know how to solve; the linker can't find some symbols (`udef_new`, `udev_enumerate_new`, etc) and I'm not sure how to fix it.

I'm not sure if I'm missing something, or if the `--json` option of kmake is missing something.

**Reply by RobDangerous** [reply]
Hm, Kinc's kfile does do an addLib('udev') call and for me it also ends up in the json file. Maybe something specific to your flavor of Linux? Do regular Kinc things build successfully?

**Reply by bluesillybeard** [reply]
The problem was that Zig uses pkg-config to link libraries, however the libraries (most notable udev) in the JSON output of kmake don't seem to work properly for pkg-config, so I just disabled that and it works now. (well, other than shaders not being compiled, since I haven't done that part yet)
